---
title: "RNAseqDemo"
author: "Matt Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 5
        number_sections: false
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      error = TRUE,
                      cache = TRUE)
```

```{r libraries, eval=TRUE}
library(edgeR)
library(tidyverse)
theme_set(theme_bw())
theme_update(plot.title = element_text(hjust = 0.5))
library(qvalue)
```

## Make up directory structure
```{bash mkdirs, eval=TRUE}
for directoryName in \
    misc \
    slurmOut \
    input/fastqs \
    output/fastqc \
    output/multiqc \
    output/align \
    output/geneCounts \
    output/figures \
    output/rdata \
    output/de \
    output/gsea
do
  if [ ! -d ${directoryName} ]
  then
    mkdir -p ${directoryName}
  fi
done
```

# Overview

This is a demo of an RNAseq analysis. We are analyzing data from this paper:
"Nrf2 contributes to the weight gain of mice during space travel"
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7479603/

The data are available on GEO and SRA:
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152382
https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA639142

I downloaded metadata including the SRA run accessions and saved it as a CSV named "misc/MiiiceIiiiinSpaaaaaaaaaaaaace.csv".

We can use this to download the raw fastq data from SRA that we'll use through the rest of this demo. We will also use the metadata during the analysis to define groups.

# Samples

We are going to analyze the RNAseq data from the thymus of four groups of mice.

- Boring wildtype C57BL/6 mice left on earth
- Terrestrial mutant Nrf2 knockout mice left on earth
- C57BL/6 astronaut mice launched into orbit for 31 days
- Mutant mice from outspace with Nrf2 knocked out launched into orbit for 31 days

# Analyses

## Get the data

I'm using prefetch and fasterq-dump to get the raw fastq files

```{bash getData, eval=TRUE}
sbatch sbatchCmds/getSRAData.sh
```

## Quality control

### Look at some of the data directly

```{bash sneakPeek, dependson='getData', eval=TRUE}
zcat input/fastqs/SRR12005052.fastq.gz | head -n 40
```


### Good fastq
```
@DB9GZKS1:667:HL3FCBCXY:1:1101:1432:2037 1:N:0:ACTTGA
CCTTGATATTTGAAGGGTCTTGGTCGGATGAAAATTTAAGCCTTTACAAGAGCAGCAAACTCTTCAGCTCCGATCTTGCCGTCACCATCGCTGTCTCCGGC
+
DDDDDIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIHIHIHIIIIIIHIIIHIIIIIIIIID
@DB9GZKS1:667:HL3FCBCXY:1:1101:1368:2106 1:N:0:ACTTGA
TATGGTTCTTTGTGAGGACGAGTTTGCAGGGCATTCATAAATAAATATAAAAAAGATGATAAGCAGAGCTCTGGGTCCGAAGGCAAGGGTCCATGAATGCT
+
DBDDDIIIIIIIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHIHIIIIIIIIIIIIIIIIIIHIIIIHHHIIIIIIIIIHH
@DB9GZKS1:667:HL3FCBCXY:1:1101:1309:2119 1:N:0:ACTTGA
CTCCAGTACACCGTTACACCTCAGCTGGTGGATAACCAGGAAGTTCTCCATAAGACCTGGAGTCTTGGACTCGTTGGGAATCAGACAACGCACAAAGTGAG
+
DDDDDIIIIIIHHHIIIIIIIIIIIIIIIIIIIGHIIIIIHHHHIHIIIIHHGHHHIHIIIIHIHIIIHHIHIIIIIIHIIHIIIIIIHIIIIGIIIIIHI
@DB9GZKS1:667:HL3FCBCXY:1:1101:1443:2144 1:N:0:ACTTGA
CATCTCTTAATGGCCAGCATCCTCTTTGAGCGGCAGTTAGGCTCCACGCACTCCAGCCTCAGCACGATCTTCTAGATCGGAAGAGCACACGTCTGAACTCC
+
DDDDDIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIGIIIIIIIIIIIIIIIII
@DB9GZKS1:667:HL3FCBCXY:1:1101:1494:2199 1:N:0:ACTTGA
CTTCATGCCAGAGATGTCCACTTTGGTGTTTGTGGCGATGCAGAACTTCTGGTGAGCTCTACGCAGAGGGACGCGGTTGATAGCCAGAGGACCGGTGACGA
+
DDDDDIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII@
@DB9GZKS1:667:HL3FCBCXY:1:1101:1376:2233 1:N:0:ACTTGA
GGCTTTTTCAGCATCTTAACCTTCCTGACGTAGACGTCATGTAGAGGATAGATGGATTGGCAGGCCTTTTCAATGTCCTTGCCAACACTGTCAGGAATCAG
+
DDDDDIIIHIIIIIIIIIIIIIIHHHIIIIIHIHIHIIIIIIHHHHIHIHIIIIIIIIIIIHIIIIIIIIHHHIIIIIIIIHHIIIIIIIIIIIIHHIHHI
@DB9GZKS1:667:HL3FCBCXY:1:1101:1688:2069 1:N:0:ACTTGA
GCAGTTTCAGTGGCCTCACCGACCTTCTCATAGATCTTCTTGGTCTCATTGTAGTCAAGGGAGGAATCATTGCACAGGGCGCAGATGGTGGCCAACTCAAC
+
DDDDDIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
@DB9GZKS1:667:HL3FCBCXY:1:1101:1580:2081 1:N:0:ACTTGA
CATTAATCCATCCATCCATTCTTTCATTTATTCATTCATTGATTCGTCCATCCATCCATCCATCCATCCATTCTTTCATTTATTCATTGATTCGTCCATCC
+
DDDDDIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIHIIHIIIIIIIIIIIHIIIHIIIIIIIIIIIIHIIIIHIIFEHHHGIHHHHIH
@DB9GZKS1:667:HL3FCBCXY:1:1101:1544:2106 1:N:0:ACTTGA
CTTGATCTTCATGGTGCTGGGGGCCAGAGCGGTGATCTCCTTCTGCATACGGTCAGCAATACCAGGGTACATGGTGGTACCGCCGGAGAGGACGTTGTTGG
+
DDDDDIIIIIIIHIIIGIIIIIIIIHHHHIIIIIHIIIIIIHHIIIHHIIIIIHIIIIIIIFHIIIIIIIIIIHIHEHHIIGIIHIHIIIIIIIIHHHHHH
@DB9GZKS1:667:HL3FCBCXY:1:1101:1530:2164 1:N:0:ACTTGA
CAGGAACATCGATAAGGGTCCAACTGCTGCCTCTCTCACTTTTGCTCTTGTATGTGGCACTGCTCTCTGTGATGGACGTCTGAGTTCGTACAAACTTTCCT
+
@0DB0CCGHHHIEEHEHHHICHEHHEHHHEHIIIIIGCFE@GH@CHIICEG<C<CH?<<FE?GHIECHHIIHHEC1</DGGE1CFD1EDCEC?<1CC@1FC
@DB9GZKS1:667:HL3FCBCXY:1:1101:1519:2184 1:N:0:ACTTGA
CAAAAAAAGTTTTCATATAACACAGATGGATTTTTTATAACACAAACAAATTTAGTAACAAAGCAGGTGAGCGGTAAGACCAGTTATACTGGAGATCGGAA
+
DDDDDIIIIGHHIIIIHHIIIIIIIIIIIIIIIIIIIIIIGHHHIIIIIHIHIIIIIIIIIIIGIIIIIHIIIIIIIIIIIHIIIIIIHIIGFHIHIIIII
```

### Short trimmed reads
```
@A00498:249:H3VJNDRXY:1:2101:1398:1000 1:N:0:NAGAGGCA+ATCATGTT
GTTTTACAGATGTGCACAAGTCTGACAGGCATCAGTTTGGTCTTCTTGAAAC
+
FFFFFFFF,:FF:F:FFFFFF:F::FFFFFFFFFFF:FFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:2555:1000 1:N:0:NAGAGGCA+ATCATGTT
GTGCTTGGCCTATTAAGTAGTATTAACTCACACTATCACAAGG
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:3441:1000 1:N:0:NAGAGGCA+ATCATGTT
NAACATACCATCCTCTGTTTCCTGTTCAATTTTCCATGGTCTCTTGGTGTTTTCAAACTGGTCCTTGCTAACAG
+
#:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:3911:1000 1:N:0:NAGAGGCA+ATCATGTT
NGCACGTCCTGATTTCCCAGCACCACGAGGTGCTCCAGGCTCTCGTGTCTTGTATTTTCCCCGCCCCAGCCTTGGAC
+
#FF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF
@A00498:249:H3VJNDRXY:1:2101:4218:1000 1:N:0:NAGAGGCA+ATCATGTT
NCCTCGCCCTCTCCCTCGCACGCCTTCCGTGTGGTACGGCCCACACCATTAGAGT
+
#FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:6099:1000 1:N:0:NAGAGGCA+ATCATGTT
NACTTAGCCTTGGGGATAAAAAAAAAAAAAATAAGATTGAGTCTCCATCCTAAAACAGGTCACGGCTAGTTCAAGAC
+
#FFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFF,:FFF:FFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFF:FF
@A00498:249:H3VJNDRXY:1:2101:8142:1000 1:N:0:NAGAGGCA+ATCATGTT
NAATGGTACTGAACCTACGAGTACACCGACTACGGCGGACTAATCTTCAACTCCTACATAC
+
#FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:9064:1000 1:N:0:NAGAGGCA+ATCATGTT
NCAGTGAGCTGAGATGACACCACTGCACTCCAGCCGGGGTGACAGAGTGAGACCC
+
#FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:9082:1000 1:N:0:NAGAGGCA+ATCATGTT
NAATAGGACTCTCCTTCTCTGCCATTTGCTTCTTCAGCATAGTTTTTGGTAAACCAAGTAACATCTTCAAGAGTGGTAG
+
#FFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:9769:1000 1:N:0:NAGAGGCA+ATCATGTT
NTATTATTCGGCGCATGAGCTGGAGTCCTAGGCACAGCTCTAAGC
+
#FF::FFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFF
```

### Reads with bad quality on the end of some reads
```
@A00498:249:H3VJNDRXY:1:2101:15356:1000 1:N:0:NAGAGGCA+ATCATGTT
NTCTTGGAGAGGAAGCCGGAAGCTGCCTTCAAGAACACCCCCAGCTCAGCCCCTGTCTCTTATACACATCTCCGAGCCCACGAGACAAGAGGCAATCTCGTATTCCGTAGTATGCGTTTAAAAAGGGGGGGGGGGGGGGGGGGGGGGGGGG
+
#FFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFF,FF:FFF,F,,F,,,,F,,,:,F,,:FF:,:,FFFFF:FFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:15519:1000 1:N:0:NAGAGGCA+ATCATGTT
NGGTGGAAGGTGATTTTATCGGAATGGGAGGTGATTCCTAGGGGGTTGTTTGATCCCGTTTCGTGCAAGAATAGGAGGTGGAGTGCTGCTAGGGCTGCAATAATGAAGGGCAAGATGAAGTGAAAGGTAAAGAATCGTGTGAGGGTGGGAC
+
#FFFFFF,FFFFFFFF:FFFFFF:FFFFFFF:FFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFF,F:FFFFF:FFFF:FFFFF:FFFFF:FFFFFF:FFFFFF:
@A00498:249:H3VJNDRXY:1:2101:15700:1000 1:N:0:NAGAGGCA+ATCATGTT
NGATCTTTTCTTCAAGAAGTAGCCCCTTTGCTAACTATAATAAAAATGTCTCCTGAGTCTCCTTATGTAGCTCTGTGTAGAAGAGAATCTAATACTTCCCTGTTGAGCAACCATGCAAGGTGGAACCAGGGAGTCGGTGGGAAGTTAGT
+
#FF,FFFFFFFFFFFFFFFFFFFF:FF:FF:FFFFFF:F:FFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFF:FFFFFF
@A00498:249:H3VJNDRXY:1:2101:15826:1000 1:N:0:NAGAGGCA+ATCATGTT
NGACACACAGGAGTGGCCAGGATCGTTTTTTAAGGAGGTGCATCTGAAACATGTTATGAAGCCTCCTTATCGAAGCAGCTCGGGAGGCTGGATGAAGACAGTAGTTGGGGCAACAGAGCAAATGGTGCAGACGCTGCCCAGATCCACACGG
+
#FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
@A00498:249:H3VJNDRXY:1:2101:16062:1000 1:N:0:NAGAGGCA+GTCATGTT
NCCCCCTCCATATGGAGTCTCGTTCTGGTCACCCAGGCTGGGGTGCTGTAGCACAATCTCTGCTCACTGCAACCTCCACCTCCTGGGTTCAAGTGATTCTCCTGCCTCAGCCTCCCTAGTAGCTGGGATTACAGGCGCCCGCCACCACGCC
+
#FFFFFFFFFFFFFFFFFF:FF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@A00498:249:H3VJNDRXY:1:2101:16297:1000 1:N:0:NAGAGGCA+ATCATGTT
NCTATAAGTCAATGAGAAGAGTTACAACTGGGAGACCCGAAAGAGTATGACCCTCCTGGCTTTAGGACTCAGGACAAAAACCAAACATCCATTTCCTTAATGAGGAGTTCACCTTTCCTCTATCCCCTGTCTCTTATACACATCTCCGAGC
+
#FFFFF::::FFFFFFFF:FFFFFFFFFFFF:FFFF:FFFFFFF,FFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFF:FFFFFFFF::FFFFF,FFFFFFFFFFFFFFFFFFFF::F:FFF,FFFFFF,F,:FFFF,
@A00498:249:H3VJNDRXY:1:2101:16315:1000 1:N:0:NAGAGGCA+ATCATGTT
NTACAGTCGCGCGATCTCAGCTCACTGTAACCTGCACCTCCCAGGCTCAAACAATTCTTCTTCCTCAGCCTCCCAAGTAGCTGGAATTATAGGTGCCCACCACCACACCCAACTAATTTTTGGACTTTTAGTAGAGACGGGGGTTTCATC
+
#FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:,FFFFFFFFFFFFFFFFFFFF:FFFFFFFFF:FFFFFFFFF:FFFFFF:FFFFFFFFFFF::F:FFFFFF:F,FF:FFFFFF:,FFFFF,:F:F,F
@A00498:249:H3VJNDRXY:1:2101:16423:1000 1:N:0:NAGAGGCA+ATCATGTT
NAGGAGAAGCAGGTGGTTGTGGCTATAAGAGGGTGGCACAAGTTATCCTTGCATAGGAACTGTTCAGTCTTTTGACGGGTGTTGAATATATGAACCTCTACAGGTGATAGAATTGTATGGAACTTAATAGACACATATACACAGGCAAATG
+
#FFFFFFFF:FFFFFFFFFFFFFF,::FF::FFFFFFFFFF:FFFFFFF,F:F:FFFFFFFFFFFF::FFFF:FFFFFF:F,FF:FFFFFF:FFFFFFFFFFF:FFF:FFFF::FFFFFFFFFF,F,FFFFFFFFFFFFFFFFFFF:FF::
@A00498:249:H3VJNDRXY:1:2101:16712:1000 1:N:0:NAGAGGCA+ATCATGTT
NCCCCATCTCATGCCACACGTCTTTCCCCCTCAATGACACCCCATCTCACGCCACGTGTCTTTCCCCCGCACAGACACCCCAACAAAACACAAATAACCATCACCCACCCCAACACACCAAAGGAACACACCAATAAAGACAACAGCATG
+
#FFFFFFFFFF,FFFFFFFFFFF:F:FFFF,FFFFFF:FFFFFFFFFFFF:FFFF,,,F:,FFF,FFF,FFF,,F:,F:FFF,:,,,,,,F:,F,,,,,,:,,:FFF,:F,F,::,,F,F,F,,,:,:F,:,,,,,,,,:,,,,,,,,,,
```

### Look at summary of data quality using fastqc

*fastqc*
Inputs:

    -   fastq.gz files

Outputs:

    -   A zip file/folder
        -   Text file summaries
        -   Images
        -   Html file
    -   An html file
        -   Overall summary

Options:

    -   -o Output folder
    -   -t # of threads
    -   --extract Extract zip folder

```{bash fastqc, eval=FALSE}
fastqc \
  -o output/fastqc \
  -t 5 \
  --extract \
  input/fastqs/*.fastq.gz
```

```{bash actualFastqc, dependson='getData', eval=TRUE}
sbatch sbatchCmds/fastqc.sh
```

## Align the reads to a reference genome

Pre-built references available on Franklin:
```
/reference/
     blast_database
     checkm
     diamond_database
     escherichia_coli
     greengenes
     homo_sapiens
     macaca_mulatta
     mus_musculus
     README.txt
     silva.119
     silva.bacteria
```

#### Mouse HiSat2 reference:

```
/reference/mus_musculus/mm10/ucsc_assmebly/illumina_download/Sequence/Hisat2Index/genome
```

### If you need to build a reference

*hisat2-build*
Inputs:

    -   Reference fasta file

Outputs:

    -   Several *.ht2 files

Options:

    -   -p # of threads

```{bash hisatBuild, eval=FALSE}
hisat2-build \
  -p 8 \
  /folder/path/to/reference/pandaGenome.fa \
  /folder/path/to/reference/pandaGenome
```

### Alignment using HiSat2

*hisat2*
Aligns fastq reads to the reference genome
Inputs:

    -   fastq.gz files
    -   Reference

Outputs:

    -   Sam file sorted by read name

Options:

    -   -x Reference genome
    -   -1 R1.fastq.gz
    -   -2 R2.fastq.gz
    -   -U R1.fastq.gz # Only if single ended reads - instead of -1/-2
    -   -p # of threads
    -   -k # of hits per Sequence to return
    -   --summary-file location to write a summary of the alignment

*samtools fixmate*
Marks which reads are PCR duplicates
Inputs:

    -   Sam file sorted by read name

Outputs:

    -   Sam file with additional info

*samtools sort*
Sort the output by position - needed to index and view in IGV
Inputs:

    -   Bam file

Outputs:

    -   Sorted bam file

Options:

    -   -@ # of threads
    -   -m how much memory to use per thread

*samtools markdup*
Removes PCR duplicates
Inputs:

    -   Bam file sorted by coordinates

Outputs:

    -   Bam file without PCR duplicates

*samtools view*
Converts sam output to bam
Inputs:

    -   Sam file

Outputs:

    -   Bam file

Options:

    -   -f filter the reads -f 2 filters reads mapped in proper pair
                            -F 3 filters unmapped reads
    -   -b output bam format

*samtools index*
Inputs:

    -   Bam file

Outputs:

    -   Bam index (bai) file

```{bash hisat, eval=FALSE}
hisat2 --help

hisat2 \
        -x /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/HiSat2Index/genome \
        -1 input/inputFile_R1.fastq.gz \
        -2 input/inputFile_R2.fastq.gz \
        -p 8 \
        --summary-file output/aligned/inputFile_Summary.txt \
    | samtools fixmate \
        -m - - \
    | samtools sort \
        - \
    | samtools markdup \
        -r - - \
    | samtools view \
        -f 2 \
        -b \
        - \
    > output/aligned/inputFile.bam

samtools index \
  output/aligned/inputFile.bam
```

### Alternate - align with tophat

*tophat*
Inputs:

    -   -p # of threads
    -   -o output folder

Outputs:

    -   accepted_hits.bam - Bam file of aligned reads
    -   junctions.bed     - Bed file of intron junctions
    -   insertions.bed    - Bed file of insertions
    -   deletions.bed     - Bed file of deletions

```{bash tophat, eval=FALSE}
tophat --help

tophat \
    -G /data/Indexes/Annotation/mm9.gtf \
    -p 8 \
    -o output/aligned \
    /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/HiSat2Index/genome \
    input/inputFile_R1.fastq.gz \
    input/inputFile_R2.fastq.gz
```

## Run the alignments on the cluster
Here I'm using a sbatch file to submit the job to the slurm scheduler. I'm using an array job to submit all the files for alignment at the same time to save time.

```{bash actualAlignment, dependson='getData', eval=TRUE}
sbatch sbatchCmds/align_hisat.sh
```

#### Check the percent of reads aligned to the genome

Ideally > 80%. If way off, look into what the other reads are.

### Look at alignment in IGV for sanity check

```{bash igv, eval=FALSE}
ml purge
ml load IGV/2.8.0-Java-11.0.2

igv.sh
```

`r knitr::include_graphics("misc/IGV.png")`

## Count reads per gene

*featureCounts*

Count number of reads per gene

Inputs:

    -   Bam files
    -   Gtf file of gene locations

Outputs:

    -   Table of reads per gene

Options:

    -   -T # of threads
    -   -a Gene annotation
    -   -o output file
    -   -p Paired end libraries (omit if single-ended)
    -   --countReadPairs Count read pairs not reads
    -   -s Stranded libraries or not

```{bash featureCounts, eval=FALSE}
featureCounts --help

featureCounts \
  -T 10 \
  -a /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Annotation/Genes/genes.gtf \
  -o output/geneCounts/combined.txt \
  -p \
  --countReadPairs \
  -s 0 \
  output/aligned/*.bam
```

```{bash actualFeatureCounts, dependson='actualAlignment', eval=TRUE}
sbatch sbatchCmds/featureCounts.sh
```

## Use multiqc to evaluate how things have run so far

```{bash multiqc, dependson='actualFeatureCounts', eval=TRUE}
ml purge
ml load GCC/9.3.0 \
        OpenMPI/4.0.3 \
        MultiQC/1.9-Python-3.8.2

multiqc \
    -o output/multiqc \
    output/
```

## Principal component analysis and differential expression analysis


### Read in gene counts
```{r countsData, dependson='actualFeatureCounts', eval=TRUE}
gene_expr <-
    read_tsv("output/geneCounts/geneCountTable.txt",
             show_col_types = FALSE,
             comment = "#") %>%
    rename_with(~ str_remove(.x, ".+\\/") %>%
                    str_remove(".bam"),
                starts_with("output")) %>%
    column_to_rownames("Geneid") %>%
    select(starts_with("SRR"))
```

### Prep metadata
```{r metadata, eval=TRUE}
metadata <-
    read_csv("misc/MiiiceIiiiinSpaaaaaaaaaaaaace.csv",
             show_col_types = FALSE) %>%
    select("Run",
           "CONDITION",
           "Genotype") %>%
    mutate(group = paste(CONDITION, Genotype, sep = "_")) %>%
    arrange(match(colnames(gene_expr),
                  Run)) %>%
    rename(sample = Run)

# Make sure the metadata and gene expression data are in the same order
# Should be TRUE
all(metadata$sample == colnames(gene_expr))
```

### Normalize the data so we can do PCA
```{r normAllData, dependson=c('countsData', 'metadata'), eval=TRUE}
groups <- factor(metadata$group)

dataDGE <- DGEList(counts = gene_expr, group = groups)

cpm(dataDGE) %>%
    apply(1, function(x) {
        sum(x != 0)
    }) %>%
    as_tibble() %>%
    ggplot(aes(x = value)) +
    geom_histogram()

# Check how many genes are in our dataset before filtering
dim(dataDGE$counts)

keptDf <- filterByExpr(dataDGE, min.count = 1)

dataDGE <- dataDGE[keptDf, keep.lib.sizes = FALSE]

# Check how many genes are in our dataset after filtering
# This step gets rid of really lowly expressed genes that can have highly inflated fold changes
dim(dataDGE$counts)

cpm(dataDGE) %>%
    apply(1, function(x) {
        sum(x != 0)
    }) %>%
    as_tibble() %>%
    ggplot(aes(x = value)) +
    geom_histogram()

# Generate normalization factors to scale the data across samples by read depth
dataDGE <- calcNormFactors(dataDGE)

dataDGE <- estimateDisp(dataDGE, model.matrix(~ groups))

# Get log transformed normalized counts per million
norm_counts <-
    as.data.frame(cpm(dataDGE, log = TRUE))

# Save the data out to a file
write.table(norm_counts,
            file = "output/de/normCounts.txt",
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")
```

## Run PCA with all samples
```{r pca, dependson='normAllData', eval=TRUE}
norm_counts <-
    read.delim("output/de/normCounts.txt",
               sep = "\t")
message("test")
# Calculate the mean and standard deviation of the log transformed normalized
# counts per million so we can assess heteroscadasticity
mean_sd_df <-
    apply(norm_counts, 1, function(x) {
        c(mean = mean(x),
          sd = sd(x))
    }) %>%
    t() %>%
    as.data.frame()

ggplot(mean_sd_df,
       aes(x = mean,
           y = sd)) +
    geom_point(alpha = 0.1) +
    ggtitle("Mean vs. SD of log transformed normalized counts per million")

# Run PCA, scaling the data so lowly expressed genes can contribute
pca_out <-
    prcomp(t(norm_counts))

# Add metadata to pca results
pc_plus_metadata <-
    pca_out$x[, 1:9] %>%
    as.data.frame() %>%
    rownames_to_column(var = "sample") %>%
    as_tibble() %>%
    left_join(metadata)

# Calculate the percent variance explianed by each PC
variance <-
    (pca_out$sdev)^2 / sum(pca_out$sdev^2) * 100
names(variance) <- colnames(pca_out$x)

plot(as.numeric(variance[1:20]))

# Plot of PC 1 and 2
pca_all <-
    ggplot(pc_plus_metadata,
           aes(x = PC1,
               y = PC2,
               color = group)) +
    geom_point(size = 2) +
    ggrepel::geom_text_repel(aes(label = sample),
                             size = 2,
                             color = "gray") +
    labs(x = paste0("PC1 (%var. = ",
                    round(variance[1], 1),
                    "%)"),
         y = paste0("PC2 (%var. = ",
                    round(variance[2], 1),
                    "%)"),
         title = "PCA of Mutant Mice From Outer Spaaaaaaace") +
    scale_color_brewer(palette = "Paired",
                       name = "Group")

ggsave("output/figures/allSamplePCA_1_2.png",
       plot = pca_all,
       width = 8,
       height = 6)

# Plot of PC 1, 2 and 3
pca_all <-
    ggplot(pc_plus_metadata,
           aes(x = PC1,
               y = PC2,
               color = group)) +
    geom_point(aes(size = PC3)) +
    ggrepel::geom_text_repel(aes(label = sample),
                             size = 2,
                             color = "gray") +
    labs(x = paste0("PC1 (%var. = ",
                    round(variance[1], 1),
                    "%)"),
         y = paste0("PC2 (%var. = ",
                    round(variance[2], 1),
                    "%)"),
         size = paste0("PC3 (%var. = ",
                       round(variance[3], 1),
                       "%)"),
         title = "PCA of Mutant Mice From Outer Spaaaaaaace") +
    scale_color_brewer(palette = "Paired",
                       name = "Group")

ggsave("output/figures/allSamplePCA_1_2_3.png",
       plot = pca_all,
       width = 8,
       height = 6)

top_9_pcs <-
    pc_plus_metadata %>%
    pivot_longer(cols = starts_with("PC"),
                 names_to = "pc",
                 values_to = "value") %>%
    ggplot(aes(x = pc,
               y = value,
               color = group,
               fill = group)) +
    geom_boxplot(color = "gray") +
    scale_fill_brewer(palette = "Paired",
                      name = "Group") +
    labs(x = "",
         y = "PC value",
         title = "Distribution of top 9 PCs")

ggsave("output/figures/top9pcs.png",
       plot = top_9_pcs,
       width = 8,
       height = 4)
```

## Run DE analysis
We're going to look at just the non-mutant mice for this analysis to keep things simple.
```{r runDE, dependson=c('countsData', 'metadata'), eval=TRUE}
# Filter down the metadata to just the non-mutant mice
sub_metadata <-
    metadata %>%
    filter(Genotype == "WT")

# Use the filtered metadata to subset the gene expression data
sub_gene_expr <-
    gene_expr[, match(sub_metadata$sample,
                      colnames(gene_expr))]

# Check that the metadata and gene expression data are in the same order
# Should be TRUE
all(sub_metadata$sample == colnames(sub_gene_expr))

groups <- factor(sub_metadata$group)

dataDGE <- DGEList(counts = sub_gene_expr, group = groups)

dim(dataDGE)

keptDf <- filterByExpr(dataDGE, min.count = 1)

dataDGE <- dataDGE[keptDf, keep.lib.sizes = FALSE]

dim(dataDGE$counts)

dataDGE <- calcNormFactors(dataDGE)

dataDGE <- estimateDisp(dataDGE, model.matrix(~ groups))

# Run the actual DE analysis
tested <- exactTest(dataDGE)

# Get all the results
results <- topTags(tested, n = Inf)
results$table$Gene <- rownames(results$table)

# Gather the results,metadata, group average expression values and individual values
results_df <-
    results$table %>%
    as_tibble() %>%
    mutate(signif = FDR <= 0.1 & abs(logFC) > log2(1.5)) %>%
    left_join(cpmByGroup(dataDGE) %>%
              as.data.frame() %>%
              rownames_to_column("Gene")) %>%
    left_join(cpm(dataDGE) %>%
              as.data.frame() %>%
              rownames_to_column("Gene"))

write.table(results_df,
            file = "output/de/edgeROut.txt",
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)
```

### Plotting DE results

```{r dePlots, dependson='runDE', eval=TRUE}
# Histogram of logFC to see if the values are centered around 0
logfc_hist <-
    ggplot(results_df, aes(x = logFC)) +
    geom_vline(xintercept = 0, color = "red") +
    geom_histogram(bins = 100) +
    ggtitle("Histogram of logFC")

ggsave("output/figures/logfc_hist.png",
       plot = logfc_hist,
       width = 8,
       height = 6)

# Histogram of p-values to get a sense of the distribution
pval_hist <-
    ggplot(results_df, aes(x = PValue)) +
    geom_histogram(bins = 200) +
    ggtitle("Histogram of p-values")

ggsave("output/figures/pval_hist.png",
       plot = pval_hist,
       width = 8,
       height = 6)

# Histogram of FDR values which should be similar to p-values but shifted left
fdr_hist <-
    ggplot(results_df, aes(x = FDR)) +
    geom_histogram(bins = 200) +
    ggtitle("Histogram of FDR values")

ggsave("output/figures/fdr_hist.png",
       plot = fdr_hist,
       width = 8,
       height = 6)

# Volcano plot
top_de <-
    results_df %>%
    filter(signif == TRUE) %>%
    arrange(desc(abs(logFC))) %>%
    group_by(logFC > 0) %>%
    slice_head(n = 50) %>%
    ungroup()

volcano <-
    ggplot(results_df,
           aes(x = logFC,
               y = -log10(PValue),
               color = signif)) +
    geom_point(alpha = 0.5) +
    ggrepel::geom_text_repel(data = top_de,
                             aes(label = Gene),
                             size = 1.5,
                             color = "gray") +
    ggtitle("Volcano plot\nFDR <= 0.1 denoted in blue") +
    scale_color_brewer(palette = "Paired")

ggsave("output/figures/volcano.png",
       plot = volcano,
       width = 10,
       height = 8)

# Scatterplot of group mean expression values (log10 transformed)
sample_scatterplot <-
    ggplot(results_df,
           aes(x = Flight_WT,
               y = Ground_control_WT,
               color = signif)) +
    geom_point(alpha = 0.2) +
    geom_abline(slope = 1, intercept = 0) +
    ggtitle("Scatterplot of mean CPM values") +
    scale_y_log10() +
    scale_x_log10() +
    labs(color = "Significant",
         x = "Space mouse",
         y = "Ground control to Major Tom") +
    scale_color_brewer(palette = "Paired")

ggsave("output/figures/sample_scatterplot.png",
       plot = sample_scatterplot,
       width = 8,
       height = 6)
```

## GSEA
msigdbr categories are defined here: https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp
```{r gsea, dependson='runDE', eval=TRUE}
kegg_ref <-
    msigdbr::msigdbr(species = "Mus musculus",
                     category = "C2",
                     subcategory = "CP:KEGG") %>%
    split(x = .$gene_symbol, f = .$gs_name)

gsea_input <-
    results_df %>%
    arrange(desc(logFC)) %>%
    pull(logFC, name = Gene)

gsea_out <-
    fgsea::fgseaMultilevel(kegg_ref,
                           gsea_input,
                           minSize = 15,
                           maxSize = 500,
                           nPerm = 1000) %>%
    arrange(desc(NES)) %>%
    filter(padj < 0.05)

write_tsv(gsea_out,
          file = paste0("output/gsea/",
                        item
                        ,
                        "category_",
                        cat_tib$category[i],
                        ".tsv"))

head(result)

my_fav_path <- "KEGG_MISMATCH_REPAIR"
fav_path_index <- grep(my_fav_path, result$pathway)
fav_path_genes <- result$leadingEdge[[fav_path_index]]

# Lets look at the expression of genes in this pathway
fav_path_expr <-
    norm_counts[rownames(norm_counts) %in% fav_path_genes, ] %>%
    rownames_to_column("gene") %>%
    pivot_longer(cols = starts_with("SRR"),
                 names_to = "sample",
                 values_to = "cpm") %>%
    left_join(metadata) %>%
    ggplot(aes(x = gene,
               y = cpm,
               fill = group)) +
    geom_boxplot() +
    labs(fill = "Group",
         x = "",
         y = "CPM",
         title = paste("Expression of genes in",
                       my_fav_path,
                       "pathway")) +
    scale_fill_brewer(palette = "Paired")

ggsave("output/figures/fav_path_expr.png",
       plot = fav_path_expr,
       width = 10,
       height = 6)
```
